wagtail/contrib/settings/utils.py

```diff
@@ -0,0 +1,91 @@
+from django.conf import settings
+from django.urls import reverse
+
+from wagtail.coreutils import get_supported_content_language_variant
+from wagtail.models import Locale, TranslatableMixin
+
+
+def get_locale_for(*, request=None, model=None, instance=None):
+    """
+    Helper to retrieve the locale to use depending on the context.
+    If a locale is returned, it is annotated to indicate whether
+    it's the default language when possible.
+    """
+    if instance and not model:
+        model = type(instance)
+
+    i18n_enabled = getattr(settings, "WAGTAIL_I18N_ENABLED", False)
+    if model:
+        i18n_enabled = i18n_enabled and issubclass(model, TranslatableMixin)
+
+    if not i18n_enabled:
+        return None
+
+    locales = Locale.objects.annotate_default_language()
+
+    if instance:
+        # Check if the instance's locale was already prefetched
+        # and use it if so to avoid an additional query.
+        field = model._meta.get_field("locale")
+        if field.is_cached(instance):
+            return instance.locale
+        else:
+            return locales.get(pk=instance.locale_id)
+
+    selected_locale = request.GET.get("locale") if request is not None else None
+    if selected_locale:
+        try:
+            return locales.get(language_code=selected_locale)
+        except Locale.DoesNotExist:
+            pass
+
+    # Fall back to the default language
+    return locales.get(
+        language_code=get_supported_content_language_variant(settings.LANGUAGE_CODE)
+    )
```

Comment:
```suggestion
    return Locale.get_default()
```

https://github.com/wagtail/wagtail/blob/4c7e1ea1ce4c842ce30905c6539437fdf21be6df/wagtail/models/i18n.py#L57